compiler/rustc_infer/src/infer/snapshot/mod.rs

```diff
@@ -60,24 +61,41 @@ impl<'tcx> InferCtxt<'tcx> {
     pub fn commit_if_ok<T, E, F>(&self, f: F) -> Result<T, E>
     where
         F: FnOnce(&CombinedSnapshot<'tcx>) -> Result<T, E>,
+        E: NoSnapshotLeaks<'tcx>,
     {
+        let no_leaks_data = E::mk_data_snapshot_start(self);
         let snapshot = self.start_snapshot();
         let r = f(&snapshot);
         debug!("commit_if_ok() -- r.is_ok() = {}", r.is_ok());
         match r {
-            Ok(_) => {
+            Ok(value) => {
                 self.commit_from(snapshot);
+                Ok(value)
             }
-            Err(_) => {
+            Err(err) => {
+                let no_leaks_data = E::mk_data_snapshot_end(self, no_leaks_data);
                 self.rollback_to(snapshot);
+                Err(E::avoid_leaks(err, self, no_leaks_data))
             }
         }
-        r
     }
 
     /// Execute `f` then unroll any bindings it creates.
     #[instrument(skip(self, f), level = "debug")]
     pub fn probe<R, F>(&self, f: F) -> R
+    where
+        F: FnOnce(&CombinedSnapshot<'tcx>) -> R,
+        R: NoSnapshotLeaks<'tcx>,
+    {
+        let no_leaks_data = R::mk_data_snapshot_start(self);
+        let snapshot = self.start_snapshot();
+        let r = f(&snapshot);
+        let no_leaks_data = R::mk_data_snapshot_end(self, no_leaks_data);
+        self.rollback_to(snapshot);
+        R::avoid_leaks(r, self, no_leaks_data)
```

Comment:
Maybe it's sufficient to run this with debug assertions? Or is it usually odd cases that have no tests that leak infer vars?